Improve support for amending git commits

mmrwoods opened the pull request:
This really a feature suggestion supplied with working code. There are a few niggles when amending git commits which in the past put me off using committia, which I've tried to address here...

When amending a commit there was no way to explicitly tell committia to show the full diff of all changes introduced by the commit, it would show you either the new changes, if there were any (i.e. `git diff` showed some changes), or, if the verbose option was provided to git commit, fall back to showing the full diff extracted from the commit message template, otherwise do nothing. When falling back to showing the full diff extracted from the commit message template, the status would not match the diff, with the diff showing changes, but status showing no changes.

This PR introduces a new opt-in feature to always extract status and diff from the commit message template when the verbose option is enabled for git commit, e.g. `git commit --verbose` or `git config --global commit.verbose=true`. This allows users to choose to see only the new changes when amending commits, e.g. `git commit --amend -a`, or show all changes, e.g. `git commit --amend -av`.

I find this really useful as I often want to see the entire diff when amending a commit and updating the message.

I've also added an, admittedly fugly, hack to fix the fall back issue with the status not matching the diff, and updated the regex to find the diff start line in the commit message template to allow for non-default git config core.commentChar, e.g. `;`

Let me know what you think, I can supply some screenshots explaining in more detail if I've not made myself clear above.

mmrwoods committed:
Optional improved support for amending git commits

Add opt-in feature to allow the diff and status to be extracted from the
commit message template if the verbose option is used with git commit,
e.g. git commit --verbose or git config --global commit.verbose=true

When amending commits, this allows the author of the commit message to
see all of the changes, in both the status and diff windows, if they
enabled the verbose option for git commit. Without the verbose option,
the behaviour is as before, show only new changes, if there are any.

Enable in your vimrc with:

    let g:committia#git#use_verbose = 1

mmrwoods committed:
Improve regex to find git diff start line

Allow for non-default git config core.commentChar, e.g. ;

Supported comment chars extracted from pattern in gitcommit syntax file,
see https://github.com/tpope/vim-git/blob/master/syntax/gitcommit.vim

mmrwoods committed:
Fix regex to extract status from commit template

mmrwoods committed:
Fix inaccurate status when amending commits

When amending commits without g:committia#git#use_verbose enabled but
with the verbose option provided to git commit, committia falls back
to extracting the diff from the commit message template when the git
diff command shows no changes, but this leads to an inaccurate status
as the diff show existing changes, but the status shows no changes.

mmrwoods commented:
I should probably point out... this code should also work when git is using a language other than English for its user interface, e.g. `LANG= ko_KR`

I've tested with `ko_KR`, `es_ES`, `zh_CN`, and `zh_TW` and all look fine to me (though I only understand the Spanish)

mmrwoods committed:
Improve code to extract status from commit template

Search backwards from the scissors line to find the start of the status
in case someone copies the same text somewhere into the commit message.

mmrwoods commented:
HI @rhysd, just wondering if you've had a chance to glance at this. Let me know if you have any questions or you want me to explain the changes in a little more detail (I can pull together some before and after screenshots if that helps).

mmrwoods committed:
Minor changes to code to extract verbose status

- Rename status_end variable to scissors_line, more accurate
- Switch single line if to multi line for readability & consistency

rhysd commented on autoload/committia/git.vim:
```diff
@@ -170,11 +181,16 @@ function! committia#git#diff() abort
 
 function! s:diff_start_line() abort
-    let re_start_diff_line = '# -\+ >8 -\+\n\%(#.*\n\)\+diff --git'
+    let re_start_diff_line = '^[#;@!$%^&|:] -\+ >8 -\+\n\%([#;@!$%^&|:].*\n\)\+diff --git'
```
Quetion: When the characters other than `#` are used at the head of line?

mmrwoods replied:
Yes, which happens when `core.commentChar` is set to something other than `#` in git config. You can test this using `git -c core.commentChar=% commit --amend --verbose`. Before this change committia will not find the scissors line and create splits for diff and status windows, after the change it works as expected.

rhysd replied:
Thanks. I didn't notice the `core.commentChar` config.

mmrwoods replied:
Actually, this doesn't work quite as expected even with this change, as `committia#git#end_of_edit_region_line()` also assumes `core.commentChar` is `#`. I'll have a look at fixing that too.

mmrwoods replied:
Fixed `committia#git#end_of_edit_region_line()` in 024a5c77

rhysd commented on autoload/committia/git.vim:
```diff
@@ -185,6 +201,24 @@ function! committia#git#status() abort
         return ''
     endtry
+    if g:committia#git#use_verbose || exists('s:use_verbose_status')
+        if exists('s:use_verbose_status') | unlet s:use_verbose_status | end
```
```suggestion
        if exists('s:use_verbose_status')
            unlet s:use_verbose_status
        endif
```

For consistency, please use multi-line `if` statement always.

rhysd commented on autoload/committia/git.vim:
```diff
@@ -185,6 +201,24 @@ function! committia#git#status() abort
+            " Save cursor position so we can restore it after searching
+            " backwards from the end of the status (i.e. scissors line)
+            " Search backwards to allow for non-default core.commentChar
+            let save_cursor = getcurpos()
+            call cursor(scissors_line, 1)
+            " localisation hack, find the start of the status in the commit
+            " message template using the first line of output from `git status`
+            let status_start = search(split(status, '\n')[0], 'benW')
+            call setpos('.', save_cursor)
```
Can we use `getline({lnum} [, {end}])` to search the line instead of `search()` so that we can avoid the side effect by moving the cursor?

mmrwoods replied:
Makes, sense, will change and test 👍

mmrwoods replied:
Ah, I'll look into this a little further, I'm already using `getline()` to actually get the relevant line, the reason I'm moving the cursor is to search backwards from the scissors line to find the start of the status. I'll see if I can find a way to do it without moving the cursor.

rhysd replied:
> I'll see if I can find a way to do it without moving the cursor.

I believe it's possible by getting the lines from start to the line by `getline(1, scissors_line)` and iterating them in reversed order.

mmrwoods replied:
Yeah, could do that, I was going to just search backwards, e.g.

```
            let status_start = search('^[#;@!$%^&|:]' . split(status, '\n')[0], 'benw')
            if status_start > 0 && status_start < scissors_line
```

But let me tinker with both options, iterating in reverse looking for a matching line does sound safer

mmrwoods replied:
Fix in d726416e, it does seem safer than simply searching backwards (people do strange things, and they should be allowed to)

rhysd reviewed (changes requested):
I apologize for the late review.

rhysd commented:
I haven't used the 'commit message template' feature. Would you show some document link for it?

rhysd commented:
I fixed CI on the master branch. Please rebase onto or merge the branch to make CI work.

mmrwoods commented:
> I haven't used the 'commit message template' feature. Would you show some document link (or `man` command line) for it?

Probably my bad choice of terminology here, the manpage for `git-commit` refers to "the commit message template" in relation to multiple options, e.g. `--no-status`. I just mean the standard template used by the `git-commit` command to generate the `COMMIT_EDITMSG` file when using an editor to prepare the commit message. 

As the `git-commit` used this terminology, I followed suit, but it is admittedly confusing as, AFAIK, this is not the same as the `-t/--template` `git-commit` option or the `commit.template` git config option.

mmrwoods commented:
> I apologize for the late review.

No problem at all, we all have lives to live, thanks for all the work on committa, I'm really enjoying using it :-)

mmrwoods committed:
Improve code to clean comments from message window

Use match() with regex to allow for non-default git `core.commentChar`

mmrwoods committed:
Further improve code to extract status from commit

Avoid moving the cursor by simply searching backwards for pattern match

mmrwoods committed:
Revert "Further improve code to extract status from commit"

Will do this by iterating in reverse from scissors line instead, safer

This reverts commit 6b91ce88b5c4865b0a7e6c50bb43eb33e6b2ec46.

mmrwoods committed:
Further improve code to extract status from commit

Avoid moving cursor by iterating in reverse from scissors line

This is safer than simply searching backwards as it allows for weirdness
like someone managing to put exactly the start of the status in code
that gets shown in the diff (yeah, I know, unlikely, but you never know)

mmrwoods committed:
Update README, document new use_verbose option

mmrwoods committed:
Safer code to clean comments from message window

Do not assume any lines starting with common comment chars is a comment,
actually get the comment char from the first line and then match that.

Note: I did initially use `s:execute_git('config core.commentChar)`, and
it worked fine, but concluded it was unnecessary, at least for now.

rhysd commented on autoload/committia/git.vim:
```diff
@@ -185,6 +201,29 @@ function! committia#git#status() abort
+            let status_start = scissors_line
+            let re_status_start = '^[#;@!$%^&|:] ' . split(status, '\n')[0]
+            while status_start > 1
+                if match(getline(status_start - 1), re_status_start) != 0
```
```suggestion
                if getline(status_start - 1) !~# re_status_start
```

`match()` changes its behavior looking at `'ignorecase'` option. It is a good idea to avoid it as much as possible.

mmrwoods replied:
Thanks, will revisit, I have force pushed since that change btw as I spotted a bug, but I'll replace any use of `match()` in another commit :-)

mmrwoods replied:
Fixed in 531ac91b, I may have gone overboard applying the change whether the pattern matches on alphabetical characters or not, but it seemed sensible to be consistent

mmrwoods committed:
Respect git comment char in status window

Prefix lines in status with git comment char rather than #

For most people, who accept the default comment char of #, this makes
absolutely no difference. However, if you do set core.commentChar to
something other than the default, this change means that character is
now consistently used in the status window, whether it is populated by
by running `git status` or extracted from the verbose commit message.

Technical note: this assumes that the last line in the message before
the diff is a comment, which it should be unless the --no-status option
is used, in which case it falls back to using '#' rather than execute a
git command. The alternative would be to run `git config core.commentChar`

mmrwoods committed:
Improve regex conditional statements, match case

Replace match() with case sensitive regex comparison operators

Apply same change to all regex conditional statements for consistency
(even ones which currently do not match any alphabetical characters)

Co-authored-by: Linda_pp <rhysd@users.noreply.github.com>

rhysd commented on autoload/committia/git.vim:
```diff
@@ -197,7 +248,7 @@ function! committia#git#end_of_edit_region_line() abort
     endif
     while line > 1
-        if stridx(getline(line - 1), '#') != 0
+        if getline(line - 1) !~# '^' . s:comment_char()
```
This doesn't work if `s:comment_char()` returns `$` because it requires escaping like `\$`. I recommend using `stridx` here.

```suggestion
        if stridx(getline(line - 1), s:comment_char()) != 0
```

mmrwoods replied:
Good spot, thanks, and the pattern match is no longer needed here as it is again matching on a single character.

mmrwoods committed:
Fix for regex special chars as comment char

Revert to using stridx() to match comment char, which avoids issues with
regex special chars being used as comment char, e.g. $. The regex match
is no longer needed as we are again matching on a single comment char.

Co-authored-by: Linda_pp <rhysd@users.noreply.github.com>

rhysd reviewed (approved):
Thanks. Looks good.

mmrwoods commented:
Thanks for merging :-)
